rama-dns/src/chain.rs

```diff
@@ -0,0 +1,236 @@
+use std::net::{Ipv4Addr, Ipv6Addr};
+
+use rama_net::address::Domain;
+
+use crate::{DenyAllDns, DnsResolver, InMemoryDns};
+
+#[derive(Debug)]
+pub struct DnsChainDomainResolveErr<E: 'static> {
+    errors: Vec<E>,
+}
+
+impl<E: std::fmt::Debug> std::fmt::Display for DnsChainDomainResolveErr<E> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "domain resolver chain resulted in errors: {:?}",
+            self.errors
+        )
+    }
+}
+
+impl<E: std::error::Error + 'static> std::error::Error for DnsChainDomainResolveErr<E> {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.errors
+            .last()
+            .map(|e| -> &(dyn std::error::Error + 'static) { e })
+    }
+}
+
+macro_rules! dns_resolver_chain_impl {
+    () => {
+        async fn ipv4_lookup(&self, domain: Domain) -> Result<Vec<Ipv4Addr>, Self::Error> {
+            let mut errors = Vec::new();
+            for resolver in self {
+                match resolver.ipv4_lookup(domain.clone()).await {
+                    Ok(ipv4s) => return Ok(ipv4s),
+                    Err(err) => errors.push(err),
+                }
+            }
+            Err(DnsChainDomainResolveErr { errors })
+        }
+
+        async fn ipv6_lookup(&self, domain: Domain) -> Result<Vec<Ipv6Addr>, Self::Error> {
+            let mut errors = Vec::new();
+            for resolver in self {
+                match resolver.ipv6_lookup(domain.clone()).await {
+                    Ok(ipv6s) => return Ok(ipv6s),
+                    Err(err) => errors.push(err),
+                }
+            }
+            Err(DnsChainDomainResolveErr { errors })
+        }
+    };
+}
+
+impl<R, E> DnsResolver for Vec<R>
+where
+    R: DnsResolver<Error = E> + Send,
+    E: Send + 'static,
+{
+    type Error = DnsChainDomainResolveErr<E>;
+
+    dns_resolver_chain_impl!();
+}
+
+impl<R, E, const N: usize> DnsResolver for [R; N]
+where
+    R: DnsResolver<Error = E> + Send,
+    E: Send + 'static,
+{
+    type Error = DnsChainDomainResolveErr<E>;
+    dns_resolver_chain_impl!();
+}
+
+#[derive(Debug)]
+enum DnsResolverType {
```

Comment:
Why did you add this type? You can use the `Either` wrapper type for this purpose, no?